### ISSUE TITLE
Symbol’s function definition is void: reverse-im-activate

### ISSUE THREAD

**Issue by jeeruff** [a13/reverse-im.el#14]
hi a13,

first of all, thanks a lot!

doom emacs version 26.2
I tried to get it work with use-package!, after installing it via doom package.el,
this in config.org:

`(use-package! reverse-im`
  `:custom`
  `(reverse-im-input-methods '("russian-computer"))`
  `:config`
  `(reverse-im-mode t))`

but it keeps giving that error:

> Symbol’s function definition is void: reverse-im-activate

I am not even sure that issue should be here and not in the doom/emacs git.
Tried to purge the packages and get it installed manually, did'nt work: the same error.

Thanks in advance and, 
yay, my first issue ever open.

**Comment by evjava** [comment]
I had the same problem with almost the same version (emacs 26.3 + doom).

I tried "the old way" and it helps:
`(use-package! reverse-im
  :ensure t
  :config
  (reverse-im-activate "russian-computer"))`

**Comment by a13** [comment]
@jeeruff  @evjava 

does it work with `use-package`?

**Comment by a13** [comment]
besides that, could you please `M-x toggle-debug-on-error` and try to eval the `use-package!` block again to see the backtrace?

**Comment by a13** [comment]
also you can `M-x pp-macroexpand-last-sexp` after the expression to see, what it expands to

**Comment by evjava** [comment]
> does it work with use-package?

Yes.

**Comment by jeeruff** [comment]
> besides that, could you please `M-x toggle-debug-on-error` and try to eval the `use-package!` block again to see the backtrace?

Debugger entered--Lisp error: (void-variable isp)
  #<subr eval-region>(936 940 nil nil)
  apply(#<subr eval-region> (936 940))
  eval-region(936 940)  ; Reading at buffer position 936
  funcall-interactively(eval-region 936 940)
  call-interactively(eval-region record nil)
  command-execute(eval-region record)
  counsel-M-x-action("eval-region")
  ivy-call()
  ivy-read("M-x " ("reverse-im-mode" "toggle-debug-on-error" "recentf-mode" "flycheck-buffer" "lisp-mode" "check-parens" "dired-narrow" "global-mu4e-conversation-mode" "reverse-im-add-input-method" "eww" "benchmark-init/show-durations-tree" "customize" "org-super-agenda-mode" "benchmark-init/show-durations-tabulated" "doom-modeline-mode" "reverse-im-which-key-show" "eval-buffer" "habitica-tasks" "all-the-icons-ibuffer-mode" "benchmark" "org-store-link" "set-input-method" "mu4e-conversation" "toggle-input-method" "mu4e-view-toggle-html" "org-dashboard-display" "which-key-show-keymap" "=rss" "org-mode" "evil-mode" "list-tags" "comment-line" "enable-theme" "counsel-faces" "writegood-mode" "mml-attach-file" "package-install" "counsel-load-theme" "customize-variable" "doom/help-packages" "doom/reload-packages" "dired-isearch-filenames" "org-mu4e-compose-org-mode" "org-agenda-toggle-deadlines" "ispell-hunspell-add-multi-dic" "org-clone-subtree-with-time-shift" "irc" "w3m" "debug" "telega" ...) :predicate #f(compiled-function (x) #<bytecode 0x43cbe62d>) :require-match t :history counsel-M-x-history :action counsel-M-x-action :keymap (keymap (67108908 . counsel--info-lookup-symbol) (67108910 . counsel-find-symbol)) :initial-input nil :caller counsel-M-x)
  counsel-M-x()
  funcall-interactively(counsel-M-x)
  call-interactively(counsel-M-x nil nil)
  command-execute(counsel-M-x)

**Comment by jeeruff** [comment]
> also you can `M-x pp-macroexpand-last-sexp` after the expression to see, what it expands to
```

(use-package reverse-im :ensure t :config
  (reverse-im-activate "russian-computer"))

```

**Comment by jeeruff** [comment]
> > does it work with use-package?
not for me, obviously.

**Comment by a13** [comment]
> > does it work with use-package?
> 
> Yes.
 
Then it looks like it's a `use-package!` problem. To be fair - I don't know why it exists at all since all the functionality can be done in use-package itself.

**Comment by a13** [comment]
> > > does it work with use-package?
> > > not for me, obviously.

Why is it obvious?

**Comment by jeeruff** [comment]
> > > > does it work with use-package?
> > > > not for me, obviously.
> 
> Why is it obvious?

hey, no offense at all. its obvious, because I am still posting and trying to find out how to solve the problem, in order to get this awesome functional tool work well with my doomacs. Let me try again on doom's discord. thanks and best wishes!

**Comment by a13** [comment]
> its obvious, because I am still posting and trying to find out how to solve the problem

I see :)

**Comment by dinkonin** [comment]
I have the same problem, reverting `(reverse-im-activate value)` back to `(mapc #'reverse-im-activate value))` in `defcustom reverse-im-input-methods` fixes the error.

> Symbol’s function definition is void: reverse-im-activate

**Comment by a13** [comment]
@dinkonin it looks like I fixed the issue, be sure you have the newest version installed (MELPA build should be ready in a couple of hours, I think). The problem was I used :custom-update, and it sets the value _after_ the library is loaded, while :custom does before. I moved reverse-im-input-methods declaration after all needed functions are defined. 

NB. You have to put `reverse-im-mode` in `:init` section instead of `:config` now.

**Comment by dinkonin** [comment]
I can confirm it works now , but strangely if I put `(reverse-im-mode t)` in :init section it does not load. It loads fine in :config so I'm leaving it there.

Thank you!

**Comment by a13** [comment]
@dinkonin it sounds strange since it doesn't load for me in :config, could you please share your full use-package snippet?

**Comment by dinkonin** [comment]
This is my config
```elisp
(use-package reverse-im
  :config
  (setq reverse-im-input-methods '("bulgarian-phonetic"))
  (require 'char-fold)
  (reverse-im-mode t))
```
Had to (require 'char-fold) because I was getting `(void-variable char-fold--default-include)` when 
doing this  ```elisp
(use-package char-fold
  :after reverse-im
  :custom
  (char-fold-include (append char-fold--default-include
                             (reverse-im-char-fold-include)))
  (char-fold-symmetric t)
  (search-default-mode #'char-fold-to-regexp))```

**Comment by a13** [comment]
I see that's because you are using setq instead of :custom. Ok if it works for you that way then.

Speaking of char-fold, you can get cyclic dependencies this way. To avoid that you can either setq char-fold-include in :config section or don't use char-fold--default-include at all 
```el
:custom
  (char-fold-include (reverse-im-char-fold-include))
```

or use my package use-package-custom-update, which sets customization after the package is loaded

```el
:custom-update
(char-fold-include (reverse-im-char-fold-include))
```

**Comment by a13** [comment]
It looks like the simplest solution is to not use `reverse-im-input-methods` :set at all, since it implicitly activates mode even when it isn't active yet (and brings a lot of other problems).

The only advantage it gives is when you are customizing its value interactively the first time. 


Sorry guys for pushing you to live on the bleeding edge…

**Comment by dinkonin** [comment]
Confirming that 
```elisp
(use-package char-fold
  :after reverse-im
  :config
  (setq char-fold-include (append char-fold--default-include (reverse-im-char-fold-include)))
  :custom
  (char-fold-symmetric t)
  (search-default-mode #'char-fold-to-regexp))
```
works without `(require 'char-fold) `

Thanks again!

**Comment by a13** [comment]
I think I'm going to introduce a defcustom to decide, whether reverse-im-mode should activate char-folding or not. 
@dinkonin

**Comment by a13** [comment]
see updated documentation

**Comment by a13** [comment]
@jeeruff  it looks like the issue fixed, feel free to reopen the issue if it doesn't work for you

Sorry for the inconvenience

**Comment by jeeruff** [comment]
--=-=-=
Content-Type: multipart/alternative; boundary="==-=-="

--==-=-=
Content-Type: text/plain; charset=utf-8; format=flowed
Content-Transfer-Encoding: quoted-printable

I can confirm it is working for me after reinstalling the 
package=20
and with this in the config.org:
```
(use-package! reverse-im
:config
(setq reverse-im-input-methods '("russian-computer"))
(use-package! char-fold
  :after reverse-im
  :config
; (setq char-fold-include (append char-fold--default-include=20
  (reverse-im-char-fold-include)))
  :custom
  (char-fold-symmetric t)
  (search-default-mode #'char-fold-to-regexp))

(reverse-im-mode t))
```

sadly, mu4e recognizes only the first key reversed, so typing 
"bt"=20
(bookmark today)
ends with b (ok) =D0=B5 (russian one). I am sure, I am being 
noob=20
commenting out that
setq char-fold line, but thats the only way it worked somehow.
Thanks a lot!
p.s. btw sending this from mu4e, I hope it will be not too 
messed=20
up. =3D)

D.K. <notifications@github.com> writes:

> @jeeruff  it looks like the issue fixed, feel free to reopen 
> if=20
> it doesn't work for you
>
> Sorry for the inconvenience


--==-=-=
Content-Type: multipart/related; boundary="===-=-="

--===-=-=
Content-Type: text/html; charset=utf-8
Content-Transfer-Encoding: quoted-printable

<p>
I can confirm it is working for me after reinstalling the package 
and with =
this in the config.org:<br />
</p>

<div class=3D"org-src-container">
<pre class=3D"src src-emacs-lisp"><span style=3D"color: 
#DCDCCC;">(</span><=
span style=3D"color: #F0DFAF; font-weight: 
bold;">use-package!</span> rever=
se-im
<span style=3D"color: #DCDCCC; font-weight: bold;">:config</span>
<span style=3D"color: #BFEBBF;">(</span><span style=3D"color: 
#F0DFAF; font=
-weight: bold;">setq</span> <span style=3D"color: 
 #DFAF8F;">reverse-im-inpu=
t-methods</span> <span style=3D"color: #51afef;">'</span><span 
style=3D"col=
or: #D0BF8F;">(</span><span style=3D"color: 
#CC9393;">"russian-computer"</s=
pan><span style=3D"color: #D0BF8F;">)</span><span style=3D"color: 
#BFEBBF;"=
>)</span>
<span style=3D"color: #BFEBBF;">(</span><span style=3D"color: 
#F0DFAF; font=
-weight: bold;">use-package!</span> char-fold
  <span style=3D"color: #DCDCCC; font-weight: bold;">:after</span> 
  reverse-=
im
  <span style=3D"color: #DCDCCC; font-weight: 
  bold;">:config</span>
  <span style=3D"color: #5F7F5F;">;</span><span style=3D"color: 
  #7F9F7F;">(=
setq char-fold-include (append char-fold--default-include 
(reverse-im-char-=
f</span><span style=3D"color: #ffffff; background-color: 
brightblack;">o</s=
pan><span style=3D"color: #7F9F7F;">ld-include)))</span>
  <span style=3D"color: #DCDCCC; font-weight: 
  bold;">:custom</span>
  <span style=3D"color: #D0BF8F;">(</span>char-fold-symmetric 
  t<span style=
=3D"color: #D0BF8F;">)</span>
  <span style=3D"color: #D0BF8F;">(</span><span style=3D"color: 
  #DFAF8F;">s=
earch-default-mode</span> <span style=3D"color: 
#51afef;">#'</span><span st=
yle=3D"color: #ECBE7B;">char-fold-to-regexp</span><span 
style=3D"color: #D0=
BF8F;">)</span><span style=3D"color: #BFEBBF;">)</span>

<span style=3D"color: #BFEBBF;">(</span><span style=3D"color: 
#DFAF8F;">rev=
erse-im-mode</span> t<span style=3D"color: #BFEBBF;">)</span><span 
style=3D=
"color: #DCDCCC;">)</span>
</pre>
</div>

<p>
sadly, mu4e recognizes only the first key reversed, so going bt 
(todays<br =
/>
messages)<br />
ends with b (ok) =D0=B5 (russian one). I am sure, I am being noob 
commentin=
g that<br />
setq, but thats the only way it worked somehow.<br />
Thanks a lot!<br />
D.K. &lt;notifications@github.com&gt; writes:<br />
</p>

<p>
&gt; @jeeruff  it looks like the issue fixed, feel free to reopen 
if it doe=
sn&rsquo;t work for you<br />
&gt;<br />
&gt; Sorry for the inconvenience<br />
</p>

--===-=-=--

--==-=-=--

--=-=-=
Content-Type: multipart/related; boundary="====-=-="

--====-=-=
Content-Type: text/html; charset=utf-8
Content-Transfer-Encoding: quoted-printable

<p>
I can confirm it is working for me after reinstalling the package 
and with =
this in the config.org:<br />
</p>

<div class=3D"org-src-container">
<pre class=3D"src src-emacs-lisp"><span style=3D"color: 
#DCDCCC;">(</span><=
span style=3D"color: #F0DFAF; font-weight: 
bold;">use-package!</span> rever=
se-im
<span style=3D"color: #DCDCCC; font-weight: bold;">:config</span>
<span style=3D"color: #BFEBBF;">(</span><span style=3D"color: 
#F0DFAF; font=
-weight: bold;">setq</span> <span style=3D"color: 
 #DFAF8F;">reverse-im-inpu=
t-methods</span> <span style=3D"color: #51afef;">'</span><span 
style=3D"col=
or: #D0BF8F;">(</span><span style=3D"color: 
#CC9393;">"russian-computer"</s=
pan><span style=3D"color: #D0BF8F;">)</span><span style=3D"color: 
#BFEBBF;"=
>)</span>
<span style=3D"color: #BFEBBF;">(</span><span style=3D"color: 
#F0DFAF; font=
-weight: bold;">use-package!</span> char-fold
  <span style=3D"color: #DCDCCC; font-weight: bold;">:after</span> 
  reverse-=
im
  <span style=3D"color: #DCDCCC; font-weight: 
  bold;">:config</span>
  <span style=3D"color: #5F7F5F;">;</span><span style=3D"color: 
  #7F9F7F;">(=
setq char-fold-include (append char-fold--default-include 
(reverse-im-char-=
f</span><span style=3D"color: #ffffff; background-color: 
brightblack;">o</s=
pan><span style=3D"color: #7F9F7F;">ld-include)))</span>
  <span style=3D"color: #DCDCCC; font-weight: 
  bold;">:custom</span>
  <span style=3D"color: #D0BF8F;">(</span>char-fold-symmetric 
  t<span style=
=3D"color: #D0BF8F;">)</span>
  <span style=3D"color: #D0BF8F;">(</span><span style=3D"color: 
  #DFAF8F;">s=
earch-default-mode</span> <span style=3D"color: 
#51afef;">#'</span><span st=
yle=3D"color: #ECBE7B;">char-fold-to-regexp</span><span 
style=3D"color: #D0=
BF8F;">)</span><span style=3D"color: #BFEBBF;">)</span>

<span style=3D"color: #BFEBBF;">(</span><span style=3D"color: 
#DFAF8F;">rev=
erse-im-mode</span> t<span style=3D"color: #BFEBBF;">)</span><span 
style=3D=
"color: #DCDCCC;">)</span>
</pre>
</div>

<p>
sadly, mu4e recognizes only the first key reversed, so going bt 
(todays<br =
/>
messages)<br />
ends with b (ok) =D0=B5 (russian one). I am sure, I am being noob 
commentin=
g that<br />
setq, but thats the only way it worked somehow.<br />
Thanks a lot!<br />
p.s. btw sending this from mu4e, I hope it will be not too messed 
up. =3D)<=
br />
</p>

<p>
D.K. &lt;notifications@github.com&gt; writes:<br />
</p>

<p>
&gt; @jeeruff  it looks like the issue fixed, feel free to reopen 
if it doe=
sn&rsquo;t work for you<br />
&gt;<br />
&gt; Sorry for the inconvenience<br />
</p>

<p>
&lt;#multipart type=3Drelated&gt;&lt;#part 
type=3Dtext/html&gt;&lt;p&gt;<br=
 />
I can confirm it is working for me after reinstalling the package 
and with =
this in the config.org:&lt;br /&gt;<br />
&lt;/p&gt;<br />
</p>

<p>
&lt;div class=3D&ldquo;org-src-container&rdquo;&gt;<br />
&lt;pre class=3D&ldquo;src src-emacs-lisp&rdquo;&gt;&lt;span 
style=3D&ldquo=
;color: #DCDCCC;&rdquo;&gt;(&lt;/span&gt;&lt;span 
 style=3D&ldquo;color: #F0=
DFAF; font-weight: bold;&rdquo;&gt;use-package!&lt;/span&gt; 
reverse-im<br =
/>
&lt;span style=3D&ldquo;color: #DCDCCC; font-weight: 
bold;&rdquo;&gt;:confi=
g&lt;/span&gt;<br />
&lt;span style=3D&ldquo;color: 
#BFEBBF;&rdquo;&gt;(&lt;/span&gt;&lt;span st=
yle=3D&ldquo;color: #F0DFAF; font-weight: 
bold;&rdquo;&gt;setq&lt;/span&gt;=
 &lt;span style=3D&ldquo;color: 
 #DFAF8F;&rdquo;&gt;reverse-im-input-methods=
&lt;/span&gt; &lt;span style=3D&ldquo;color: 
#51afef;&rdquo;&gt;&rsquo;&lt;=
/span&gt;&lt;span style=3D&ldquo;color: 
#D0BF8F;&rdquo;&gt;(&lt;/span&gt;&l=
t;span style=3D&ldquo;color: 
#CC9393;&rdquo;&gt;&ldquo;russian-computer&rdq=
uo;&lt;/span&gt;&lt;span style=3D&ldquo;color: 
#D0BF8F;&rdquo;&gt;)&lt;/spa=
n&gt;&lt;span style=3D&ldquo;color: 
#BFEBBF;&rdquo;&gt;)&lt;/span&gt;<br />
&lt;span style=3D&ldquo;color: 
#BFEBBF;&rdquo;&gt;(&lt;/span&gt;&lt;span st=
yle=3D&ldquo;color: #F0DFAF; font-weight: 
bold;&rdquo;&gt;use-package!&lt;/=
span&gt; char-fold<br />
  &lt;span style=3D&ldquo;color: #DCDCCC; font-weight: 
  bold;&rdquo;&gt;:aft=
er&lt;/span&gt; reverse-im<br />
  &lt;span style=3D&ldquo;color: #DCDCCC; font-weight: 
  bold;&rdquo;&gt;:con=
fig&lt;/span&gt;<br />
  &lt;span style=3D&ldquo;color: 
  #5F7F5F;&rdquo;&gt;;&lt;/span&gt;&lt;span =
style=3D&ldquo;color: #7F9F7F;&rdquo;&gt;(setq char-fold-include 
(append ch=
ar-fold&#x2013;default-include 
(reverse-im-char-f&lt;/span&gt;&lt;span styl=
e=3D&ldquo;color: #ffffff; background-color: 
brightblack;&rdquo;&gt;o&lt;/s=
pan&gt;&lt;span style=3D&ldquo;color: 
#7F9F7F;&rdquo;&gt;ld-include)))&lt;/=
span&gt;<br />
  &lt;span style=3D&ldquo;color: #DCDCCC; font-weight: 
  bold;&rdquo;&gt;:cus=
tom&lt;/span&gt;<br />
  &lt;span style=3D&ldquo;color: 
  #D0BF8F;&rdquo;&gt;(&lt;/span&gt;char-fold=
-symmetric t&lt;span style=3D&ldquo;color: 
 #D0BF8F;&rdquo;&gt;)&lt;/span&gt=
;<br />
  &lt;span style=3D&ldquo;color: 
  #D0BF8F;&rdquo;&gt;(&lt;/span&gt;&lt;span =
style=3D&ldquo;color: 
#DFAF8F;&rdquo;&gt;search-default-mode&lt;/span&gt; &=
lt;span style=3D&ldquo;color: 
#51afef;&rdquo;&gt;#&rsquo;&lt;/span&gt;&lt;s=
pan style=3D&ldquo;color: 
#ECBE7B;&rdquo;&gt;char-fold-to-regexp&lt;/span&g=
t;&lt;span style=3D&ldquo;color: 
#D0BF8F;&rdquo;&gt;)&lt;/span&gt;&lt;span =
style=3D&ldquo;color: #BFEBBF;&rdquo;&gt;)&lt;/span&gt;<br />
</p>

<p>
&lt;span style=3D&ldquo;color: 
#BFEBBF;&rdquo;&gt;(&lt;/span&gt;&lt;span st=
yle=3D&ldquo;color: 
#DFAF8F;&rdquo;&gt;reverse-im-mode&lt;/span&gt; t&lt;sp=
an style=3D&ldquo;color: #BFEBBF;&rdquo;&gt;)&lt;/span&gt;&lt;span 
style=3D=
&ldquo;color: #DCDCCC;&rdquo;&gt;)&lt;/span&gt;<br />
&lt;/pre&gt;<br />
&lt;/div&gt;<br />
</p>

<p>
&lt;p&gt;<br />
sadly, mu4e recognizes only the first key reversed, so going bt 
(todays&lt;=
br /&gt;<br />
messages)&lt;br /&gt;<br />
ends with b (ok) =D0=B5 (russian one). I am sure, I am being noob 
commentin=
g that&lt;br /&gt;<br />
setq, but thats the only way it worked somehow.&lt;br /&gt;<br />
Thanks a lot!&lt;br /&gt;<br />
D.K. &amp;lt;notifications@github.com&amp;gt; writes:&lt;br 
/&gt;<br />
&lt;/p&gt;<br />
</p>

<p>
&lt;p&gt;<br />
&amp;gt; @jeeruff  it looks like the issue fixed, feel free to 
reopen if it=
 doesn&amp;rsquo;t work for you&lt;br /&gt;<br />
&amp;gt;&lt;br /&gt;<br />
&amp;gt; Sorry for the inconvenience&lt;br /&gt;<br />
&lt;/p&gt;<br />
&lt;#/multipart&gt;<br />
&lt;#/multipart&gt;<br />
</p>

--====-=-=--

--=-=-=--

**Comment by a13** [comment]
@jeeruff please read the updated README, you don't need to explicitly set up char-fold-include, set `reverse-im-char-fold` to `t` before turning on `reverse-im-mode`.

As for mu4e - I've created an issue for that!
